From 279f3461b310661bac99cf13791feab3f9b22c61 Mon Sep 17 00:00:00 2001
From: Sebastian Bazley <sebb@apache.org>
Date: Tue, 6 Jan 2015 02:37:09 +0000
Subject: [PATCH] VALIDATOR-235 UrlValidator rejects url with Unicode
 characters in domain label or TLD Add URL test and fix up domain so it works

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/validator/trunk@1649721 13f79535-47bb-0310-9956-ffa450edef68
---
 .../validator/routines/DomainValidator.java   |  3 ++-
 .../validator/routines/UrlValidator.java      | 20 ++++++++++++++++++-
 .../validator/routines/UrlValidatorTest.java  | 13 ++++++++++++
 3 files changed, 34 insertions(+), 2 deletions(-)

diff --git a/src/main/java/org/apache/commons/validator/routines/DomainValidator.java b/src/main/java/org/apache/commons/validator/routines/DomainValidator.java
index 93c739484..7b6940ecc 100644
--- a/src/main/java/org/apache/commons/validator/routines/DomainValidator.java
+++ b/src/main/java/org/apache/commons/validator/routines/DomainValidator.java
@@ -1070,7 +1070,8 @@ private String chompLeadingDot(String str) {
      * @param input the string to convert, not null
      * @return converted input, or original input if conversion fails
      */
-    private static String unicodeToASCII(String input) {
+    // Needed by UrlValidator
+    static String unicodeToASCII(String input) {
         try {
             return /* java.net.IDN. */ toASCII(input);
         } catch (IllegalArgumentException e) { // input is not valid
diff --git a/src/main/java/org/apache/commons/validator/routines/UrlValidator.java b/src/main/java/org/apache/commons/validator/routines/UrlValidator.java
index 6533eb7c9..d386ad3c5 100644
--- a/src/main/java/org/apache/commons/validator/routines/UrlValidator.java
+++ b/src/main/java/org/apache/commons/validator/routines/UrlValidator.java
@@ -160,6 +160,12 @@ public class UrlValidator implements Serializable {
     private static final String PORT_REGEX = "^:(\\d{1,5})$";
     private static final Pattern PORT_PATTERN = Pattern.compile(PORT_REGEX);
 
+    // Pattern to extract domain for IDN conversion
+    private static final Pattern HTTP_IDN_PATTERN = Pattern.compile("(https?://)([^/]+)(.*)", Pattern.CASE_INSENSITIVE);
+    private static final int PARSE_HTTP_IDN_SCHEME = 1;
+    private static final int PARSE_HTTP_IDN_AUTH = 2;
+    private static final int PARSE_HTTP_IDN_REST = 3;
+
     /**
      * Holds the set of current validation options.
      */
@@ -290,7 +296,19 @@ public boolean isValid(String value) {
         }
 
         if (!ASCII_PATTERN.matcher(value).matches()) {
-            return false;
+            // Non-ASCII input, try and convert HTTP domain
+            Matcher httpMatcher = HTTP_IDN_PATTERN.matcher(value);
+            if (httpMatcher.lookingAt()) { // We have an http(s) URL
+                value =   httpMatcher.group(PARSE_HTTP_IDN_SCHEME)
+                        + DomainValidator.unicodeToASCII(httpMatcher.group(PARSE_HTTP_IDN_AUTH)) 
+                        + httpMatcher.group(PARSE_HTTP_IDN_REST);
+                if (!ASCII_PATTERN.matcher(value).matches()) {
+                    return false;
+                }
+                // Drop thru, we were able to convert the pattern
+            } else {
+                return false;
+            }
         }
 
         // Check the whole url address structure
diff --git a/src/test/java/org/apache/commons/validator/routines/UrlValidatorTest.java b/src/test/java/org/apache/commons/validator/routines/UrlValidatorTest.java
index 5eb16722e..67f386d66 100644
--- a/src/test/java/org/apache/commons/validator/routines/UrlValidatorTest.java
+++ b/src/test/java/org/apache/commons/validator/routines/UrlValidatorTest.java
@@ -142,6 +142,19 @@ public void testValidator218() {
                validator.isValid("http://somewhere.com/pathxyz/file(1).html"));
    }
 
+   public void testValidator235() {
+       String version = System.getProperty("java.version");
+       if (version.compareTo("1.6") < 0) {
+           System.out.println("Cannot run Unicode IDN tests");
+           return; // Cannot run the test
+       }
+       UrlValidator validator = new UrlValidator();
+       assertTrue("xn--d1abbgf6aiiy.xn--p1ai should validate", validator.isValid("http://xn--d1abbgf6aiiy.xn--p1ai"));
+       assertTrue("президент.рф should validate", validator.isValid("http://президент.рф"));
+       assertTrue("www.b\u00fccher.ch should validate", validator.isValid("http://www.b\u00fccher.ch"));
+       assertFalse("www.\uFFFD.ch FFFD should fail", validator.isValid("http://www.\uFFFD.ch"));
+   }
+
     public void testValidator248() {
         RegexValidator regex = new RegexValidator(new String[] {"localhost", ".*\\.my-testing"});
         UrlValidator validator = new UrlValidator(regex, 0);
